Refactor Parser

fpringvaldsen opened the pull request:
This PR refactors the Parser to parse Config Nodes rather than Tokens. It also contains some fixes for bugs found in the ConfigDocumentParser while performing the refactoring.

havocp commented:
Travis is saying it doesn't build, with a couple compile errors to look at.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigDocumentParser.java:
```diff
@@ -106,9 +116,13 @@ private boolean checkElementSeparator(Collection<AbstractConfigNode> nodes) {
+                    } else if (Tokens.isComment(t)) {
+                        nodes.add(new ConfigNodeComment(t));
+                        t = nextToken();
+                        continue;
```
this is a little clunky to sometimes return, sometimes continue, and sometimes fall down to add ConfigNodeSingleToken.
I'd suggest putting a nodes.add in every branch, and then for those that don't return always fall down to `t = nextToken`
(i.e. remove `t = nextToken` and `continue` from this branch, and add `nodes.add(new ConfigNodeSingleToken)` to the two branches that don't continue or return).

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigDocumentParser.java:
```diff
@@ -410,9 +425,9 @@ private ConfigNodeComplexValue parseObject(boolean hadOpenCurly) {
-
+                    ArrayList<AbstractConfigNode> includeNodes = new ArrayList<AbstractConfigNode>();
+                    includeNodes.add(new ConfigNodeSingleToken(t));
+                    objectNodes.add(parseInclude(includeNodes));
```
if the tests are passing I must be confused here, but it sure looks like only the token `include` gets passed to parseInclude, and then parseInclude is parsing the filename after `include` as well (which I don't see how it has).
I'm sure I'll figure it out after I post this comment ;-)

havocp replied:
oh I completely missed that nextTokenIgnoringWhitespace is adding to that array. I get it now

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigDocumentParser.java:
```diff
@@ -72,9 +74,17 @@ private Token nextToken() {
         private Token nextTokenIgnoringWhitespace(Collection<AbstractConfigNode> nodes) {
```
let's rename this `nextTokenCollectingWhitespace` maybe - would make more sense given its current behavior.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigDocumentParser.java:
```diff
@@ -197,7 +211,7 @@ private ConfigException parseError(String message) {
 
         private ConfigException parseError(String message, Throwable cause) {
-            return new ConfigException.Parse(SimpleConfigOrigin.newSimple(""), message, cause);
+            return new ConfigException.Parse(SimpleConfigOrigin.newSimple("").withLineNumber(lineNumber), message, cause);
```
the old parser has a `baseOrigin` passed in, rather than making up this `newSimple("")` origin - I would think ConfigDocumentParser.parse should take an origin too? then you don't have to make one up here. If the `newSimple("")` origin escapes into public API, then we probably need a test that it doesn't... we shouldn't ever be returning an empty/useless origin like that.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigDocumentParser.java:
```diff
@@ -381,6 +395,7 @@ private void parseInclude(ArrayList<AbstractConfigNode> children) {
             } else {
                 throw parseError("include keyword is not followed by a quoted string, but by: " + t);
             }
+            return new ConfigNodeInclude(children);
```
the list of include types (url, file, classpath) are parsed in this parseInclude method, but then it looks like we toss the result, which makes me expect that the same switch on url/file/classpath is going to be elsewhere in the code - I think only one place in the code should contain the four strings (url, file, classpath, error) and then all other places should have something pre-parsed to work with. further down maybe some ideas on where to consolidate

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeInclude.java:
```diff
@@ -0,0 +1,31 @@
+            if (n instanceof ConfigNodeSingleToken) {
+                Token t = ((ConfigNodeSingleToken) n).token();
+                if (Tokens.isUnquotedText(t) && !t.tokenText().equals("include") &&
+                        t.tokenText().matches(".*\\w.*")) {
```
yeah - this is kind of a mini-parsing step here. I think it'd be better to grab the kind in the parser and then pass it to the ConfigNodeInclude constructor or something. Uses a bit more memory but avoids having so much knowledge of syntax here in this method after the parser has completed.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeRoot.java:
```diff
@@ -0,0 +1,38 @@
+                return (ConfigNodeComplexValue)node;
+            }
+        }
+        throw new ConfigException.BugOrBroken("ConfigNodeRoot did not contain a value");
```
does the parser check for this and throw before creating the ConfigNodeRoot? If not we could reach this exception due to an empty config file, rather than due to a bug in the program. (I'm not sure a zero-byte config file is even disallowed - it may have been ok to do `parseString("")` before - I'd sort of expect it was, but I'm not sure I ever tried it)

fpringvaldsen replied:
It does not. 

From what I can tell, if an empty document manages to make its way down to the parser, it will create an empty object and pass that to `ConfigNodeRoot`, so that should always have a value associated with it. I definitely think it's worth writing a test or two to validate this behavior.

fpringvaldsen replied:
Okay, wrote some tests, and I've verified that the ConfigDocumentParser will create an empty object in the case of an empty document.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeRoot.java:
```diff
@@ -0,0 +1,38 @@
+            AbstractConfigNode node = childrenCopy.get(i);
+            if (node instanceof ConfigNodeComplexValue) {
+                if (node instanceof ConfigNodeArray) {
+                    throw new ConfigException.Generic("The ConfigDocument had an array at the root level, and values cannot be replaced inside an array.");
```
maybe ConfigException.WrongType would make sense here? we should also mention this case in the API docs.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeSimpleValue.java:
```diff
@@ -16,4 +17,21 @@
+
+            return new ConfigReference(token.origin(), new SubstitutionExpression(path, optional));
+        }
+        return null;
```
I'm suspicious of this null - when does it happen? Often I'd call a method that can return null `valueOrNull` for clarity, but depending on when it happens I don't know if it should be an exception instead, or what.

fpringvaldsen replied:
Yeah, I don't think this `return null` should ever be reached, it should probably throw a `BugOrBrokenException`.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigParser.java:
```diff
@@ -0,0 +1,435 @@
+        // the number of lists we are inside; this is used to detect the "cannot
+        // generate a reference to a list element" problem, and once we fix that
+        // problem we should be able to get rid of this variable.
+        int arrayCount;
```
this feels like we have several things replicated between the two parsers; that feels like some kind of separation-of-concerns problem, but I have to read on before I have a constructive suggestion :-)

havocp replied:
In my skimming, both parsers track equalsCount and arrayCount, but only ConfigDocumentParser uses equalsCount and only ConfigParser uses arrayCount. So you could have only equalsCount in DocumentParser and only arrayCount in ConfigParser ?

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigParser.java:
```diff
@@ -0,0 +1,435 @@
+        private AbstractConfigValue parseConcatenation(ConfigNodeConcatenation n) {
+            // this trick is not done in JSON
+            if (flavor == ConfigSyntax.JSON)
+                return null;
```
if we're in JSON, shouldn't the DocumentParser have thrown an exception rather than creating a ConfigNodeConcatenation? If so then this `return null` would be impossible (it could be a BugOrBroken instead perhaps)

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigParser.java:
```diff
@@ -0,0 +1,435 @@
+                obj = (AbstractConfigObject) includer
+                        .include(includeContext, n.name());
+            }
+
```
Maybe an improvement to this `kind` situation would be an enum `ConfigIncludeKind`, compute the enum from the strings in DocumentParser, the enum values are `URL, FILE, CLASSPATH, HEURISTIC`, then `ConfigNodeInclude.kind` returns the enum value (never null) and we switch on the enum value here ? Then we don't have the parse logic here and in ConfigNodeInclude.kind

havocp commented on config/src/main/java/com/typesafe/config/impl/Parseable.java:
```diff
@@ -254,7 +254,9 @@ private AbstractConfigValue rawParseValue(Reader reader, ConfigOrigin origin,
-            return Parser.parse(tokens, origin, finalOptions, includeContext());
+            ConfigNodeRoot document = ConfigDocumentParser.parse(tokens, finalOptions);
+            return ConfigParser.parse(document, origin, finalOptions, includeContext());
+            //return Parser.parse(tokens, origin, finalOptions, includeContext());
```
delete this commented out line I guess

havocp commented:
I think ConfigDocumentParser's `pathStack` may never have anything pushed onto it.

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigParser.java:
```diff
@@ -0,0 +1,435 @@
+        final private FullIncluder includer;
+        final private ConfigIncludeContext includeContext;
+        final private ConfigSyntax flavor;
+        final private ConfigOrigin baseOrigin;
```
I wonder if we can track lineNumber and origin on each node, and flavor on the document, so those three can be omitted here ?

IIRC some of the tokens (such as braces) don't track line and origin in an attempt to be more efficient; but we can maybe compute line and origin by starting from a token that does have them and then scanning looking at newlines, or something.
I'm not sure exactly how to do it. I guess we could leave this as a future enhancement.

havocp commented:
Thanks! I'm glad to see this is a net decrease in lines of code, too.

If all the existing parser tests are passing I think we can feel pretty good about things.

fpringvaldsen commented:
@havocp Sorry about the Travis failure! I was running all the tests in an IDE and that wasn't showing me the compile error, I'm fixing them now. I should be able to address your feedback tomorrow.

fpringvaldsen commented:
Thanks for the feedback as always!

fpringvaldsen commented:
@havocp I've addressed your feedback!

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigIncludeKind.java:
```diff
@@ -0,0 +1,5 @@
+package com.typesafe.config.impl;
+
+enum ConfigIncludeKind {
+    URL, FILE, CLASSPATH, HEURISTIC, STRING
```
STRING is never used I think? just a string with no `url()` etc. around it would be heuristic "do what I mean" mode.

havocp commented:
Couple more minor things then I think we can merge this. What are your plans for the next step? I guess this is everything you needed for your use-case?

We don't have a public ConfigNode API here yet... I was just thinking maybe we should move it all into a subpackage like `com.typesafe.config.parser` or `com.typesafe.config.ast` or `com.typesafe.config.hocon` so it doesn't clutter the docs for people who are just trying to read config settings. Then roughly speaking I guess I'd expect some public node subtypes and a `ConfigNodeFactory` or along those lines. It is fine if you don't want to work on the public API or it isn't in-scope for you though; it can be a future elaboration.

For this PR, I think just clean up that ConfigIncludeKind enum with the extra value, and maybe use a switch statement (which would have caught the extra value), and we can merge what you have here.

Thanks!

fpringvaldsen commented:
@havocp I've addressed your feedback!

This is about everything we need for our use case (I think the only other thing is two more methods for ConfigDocument, `hasValue()` and `removeValue()`, which I'm going to open a PR for pretty soon). I can also definitely take a swing at the indentation issue when adding a new setting, but I think that's probably all we can afford to do for now.

Thanks!

havocp commented:
Thanks! Merged. I have some stuff I want to do here (I think I'll put the public API in its own package as I mentioned, and maybe make some public ConfigNode types). But it doesn't sound like it will conflict with your plans except in superficial easily-resolved ways. I'm not sure when I'll get around to it anyhow... hopefully soon though so we can do an -M2 with this.
